File: day3/day3.py
global gamma_rate 
gamma_rate = ""
global epsilon_rate 
epsilon_rate = ""

global ceros
ceros = []
global unos
unos = []


def recorrer(data):
    global count
    global unos
    global ceros
    for i in data[1]:
        unos.append(0)
        ceros.append(0)

    for val in data:
        for i,bit in enumerate(val):
            if bit == "1":
                unos[i] += 1
            else:
                ceros[i] += 1



def mayoritario(un, ce):
    global gamma_rate
    global epsilon_rate
    for u,c in zip(un,ce):
        if u > c:
            gamma_rate += "1"
            epsilon_rate += "0"
        else:
            gamma_rate += "0"
            epsilon_rate += "1"




if __name__=="__main__":
    data = [line.strip() for line in open("input.txt","r").readlines()]
    recorrer(data)
    mayoritario(unos, ceros)

    print(int(gamma_rate,2) * int(epsilon_rate,2))

File: day3/day3.java
package day3;

import java.io.File;
import java.util.Scanner;

public class day3 {

    public static void main(String[] args) {
        // List<Integer> unos = new ArrayList<>(12);
        // List<Integer> ceros = new ArrayList<>(12);
        int[] unos = new int[12],ceros = new int[12];
        String gamma_rate = "", epsilon_rate = ""; 


        
        File fd = new File("day3/input.txt");
        Scanner file = null;
        try {
            String line = "";
            file = new Scanner(fd);

            //recorremos el archivo
            while(file.hasNext()){
                line = file.nextLine();
                //comprobamos que bit hay en cada posicion
                for(int i = 0; i<line.length();i++){
                    if(line.charAt(i) == '0'){
                        ceros[i]++;
                    }else{
                        unos[i]++;
                    }
                }
            }

            // comprobamos cual es el bit mayoritario
            for(int i=0;i<unos.length;i++){
                if(unos[i]>ceros[i]){
                    gamma_rate += "1";
                    epsilon_rate += "0";
                } else {
                    gamma_rate += "0";
                    epsilon_rate += "1";
                }
            }

            int sum = Integer.parseInt(epsilon_rate,2) * Integer.parseInt(gamma_rate,2);

            System.out.println(sum);

        } catch (Exception e) {
            System.out.println(e.toString());
        }
    }
    
}
